staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go

```diff
@@ -228,87 +281,22 @@ var newETCD3Client = func(c storagebackend.TransportConfig) (*clientv3.Client, e
 	return clientv3.New(cfg)
 }
 
-type runningCompactor struct {
-	interval time.Duration
-	cancel   context.CancelFunc
-	client   *clientv3.Client
-	refs     int
-}
-
 var (
-	// compactorsMu guards access to compactors map
-	compactorsMu sync.Mutex
-	compactors   = map[string]*runningCompactor{}
 	// dbMetricsMonitorsMu guards access to dbMetricsMonitors map
 	dbMetricsMonitorsMu sync.Mutex
 	dbMetricsMonitors   map[string]struct{}
 )
 
-// startCompactorOnce start one compactor per transport. If the interval get smaller on repeated calls, the
-// compactor is replaced. A destroy func is returned. If all destroy funcs with the same transport are called,
-// the compactor is stopped.
-func startCompactorOnce(c storagebackend.TransportConfig, interval time.Duration) (func(), error) {
-	compactorsMu.Lock()
-	defer compactorsMu.Unlock()
-
-	key := fmt.Sprintf("%v", c) // gives: {[server1 server2] keyFile certFile caFile}
-	if compactor, foundBefore := compactors[key]; !foundBefore || compactor.interval > interval {
-		compactorClient, err := newETCD3Client(c)
-		if err != nil {
-			return nil, err
-		}
-
-		if foundBefore {
-			// replace compactor
-			compactor.cancel()
-			compactor.client.Close()
-		} else {
-			// start new compactor
-			compactor = &runningCompactor{}
-			compactors[key] = compactor
-		}
-
-		ctx, cancel := context.WithCancel(context.Background())
-
-		compactor.interval = interval
-		compactor.cancel = cancel
-		compactor.client = compactorClient
-
-		etcd3.StartCompactor(ctx, compactorClient, interval)
-	}
-
-	compactors[key].refs++
-
-	return func() {
-		compactorsMu.Lock()
-		defer compactorsMu.Unlock()
-
-		compactor := compactors[key]
-		compactor.refs--
-		if compactor.refs == 0 {
-			compactor.cancel()
-			compactor.client.Close()
-			delete(compactors, key)
-		}
-	}, nil
-}
-
-func newETCD3Storage(c storagebackend.ConfigForResource, newFunc func() runtime.Object) (storage.Interface, DestroyFunc, error) {
-	stopCompactor, err := startCompactorOnce(c.Transport, c.CompactionInterval)
+func newETCD3Storage(client *etcd3ClientCache, tc storagebackend.ConfigForResource, newFunc func() runtime.Object) (storage.Interface, DestroyFunc, error) {
+	c, stopClient, err := client.For(tc.Transport, tc.CompactionInterval)
 	if err != nil {
 		return nil, nil, err
 	}
 
-	client, err := newETCD3Client(c.Transport)
-	if err != nil {
-		stopCompactor()
-		return nil, nil, err
-	}
-
 	// decorate the KV instance so we can track etcd latency per request.
-	client.KV = etcd3.NewETCDLatencyTracker(client.KV)
+	c.KV = etcd3.NewETCDLatencyTracker(c.KV)
```

Comment:
how does this will behave? 
this is wrapping latency trackers on top of the same client. right?

Reply:
I haven't looked into how the latency tracker works yet. The key things that have changed here are:

* Still one tracker per KV client, but now each KV client potentially serves more callers.
* The wrapped etcd client is also used for compaction - previously that client had no latency tracker AFAICT.

Reply:
Ah I think I get what you're saying now. We should be creating one of these per client, not per storage. This is causing a data race per the unit tests. Will fix.

Reply:
This is done in the latest iteration.